### PR TITLE
Add logic for preserving unloaded Cogs.

### PR DESCRIPTION
**Pull request by Shivansh-007**
This PR makes a Redis cache in `bot.bot` to store all the unloaded extensions, on every unload and load this cache would be edited in the desired manner, either removing it from it or adding it in the cache.

Now, coming to not loading the unloaded commands, on bot startup, just after Redis gets connected, the bot would run `bot.get_unloaded_extensions` and get all the unloaded extensions from the cache. These extensions would be skipped and not loaded, the skipping message would be logged. Once all these extensions are loaded, we compare and check which
extensions are present in the cache and not the bot. This means these cogs are either removed or renamed.

A message containing all these "extra" cogs are sent to #dev-alerts and these cogs are removed from the cache.

Resolves: #705

##### 1: Cache containing `'bot.exts.evergreen.catify | bot.exts.evergreen.xkd'`
Screenshot from 2021-05-18 15-37-50

##### 2: After bot sends ^
Screenshot from 2021-05-18 15-43-34

##### 3: On doing `.exts unload ping`
Screenshot from 2021-05-18 15-44-57

##### 4: On doing `.exts load ping`
Screenshot from 2021-05-18 15-57-27

### Todo
- [x] There is a small bug in this, I will fix that soon, since the bot removes the cogs from the cache first and then performs the action, it will also remove the cogs which have been failed to auction.
- [ ] Remove testing code
- [ ] Add core-developers role to allowed mention, need to discuss if it should be allowed overall or just while the alert message is sent.

### CODE REVIEW

File: bot/bot.py
Comment (ToxicKidz) [inline bot/bot.py:86]: Maybe we can just override `load_extension` here

File: bot/constants.py
Comment (ToxicKidz) [inline bot/constants.py:274]: Did you forget to change this back to the correct ID?

**Review by ToxicKidz — CHANGES_REQUESTED** [review]

File: bot/constants.py
Comment (Shivansh-007) [reply bot/constants.py:274]: Whoops yeah

File: bot/bot.py
Comment (Shivansh-007) [reply bot/bot.py:86]: Hmm, we could but I don’t see why is that better than this.

Comment (ToxicKidz) [reply bot/bot.py:86]: `Cog.__module__` is not the same as its extension. For example, `InternalEval.__module__` would be `bot.exts.internal_eval._internal_eval`, yet the corresponding extension is `bot.exts.internal_eval`.

Comment (Shivansh-007) [reply bot/bot.py:86]: I am using `Cog.__name__` here not `.__module__`

Screenshot from 2021-05-22 07-42-11

File: bot/bot.py
Comment (ToxicKidz) [inline bot/bot.py:72]: This should be `Optional[str]`.
```suggestion
    def load_extension(self, name: str, *, package: Optional[str] = None) -> None:
```

File: bot/bot.py
Comment (ToxicKidz) [inline bot/bot.py:216]:
```suggestion
        extensions_msg = "- " + "\n- ".join(self.unloaded_extensions)
```

**Review by ToxicKidz — CHANGES_REQUESTED** [review]

**Review by kosayoda — CHANGES_REQUESTED** [review]
From a cursory look at the PR:

I don't think this is a suitable way to store unloaded cogs in the cache. Currently, you manually serialize cog names by joining them with a pipe character, then setting it to a single key `unloaded` in the cache.

This is redundant when we have a cache to work with: Simply set the cog name as the cache key. That way, there is no need to even deserialize the cache into a dictionary on every time you want to modify the cache: Simply set and get the cog names from the cache.

That way, the value for each key in the cache is also freed up. One useful value for that is the message ID used to unload the cogs. In the event of an unloaded cog not found on startup, the embed can provide that message ID/link to give more context to core developers.

File: bot/bot.py
Comment (kosayoda) [inline bot/bot.py:50]: Are these for local testing? They should not be left in production.

File: bot/bot.py
Comment (kosayoda) [inline bot/bot.py:44]: Is this guaranteed to run after all the extensions have been loaded?

File: bot/bot.py
Comment (kosayoda) [inline bot/bot.py:218]: You should be able to get the message link for an extension straight from the rediscache without having to get the whole cache as a dictionary.

**Review by kosayoda — CHANGES_REQUESTED** [review]
With your current approach, you store the cached unloads into a dict, then remove successfully unloaded cogs from the dict in `load_extension`, then compare the dict with the cached unloads to find missing cogs.

A better solution would be to just go ahead with loading the extension (and checking presence in the cache before loading), and storing a set of successfully unloaded cogs instead. Then, you can get all the unloads from the cache and clean up the ones that were not in the successful dict.

This can be done with a `load_extensions` method on the bot that keep tracks of extensions that were successfully unloaded locally. That way, there's no need to store a redundant instance attribute after startup. It would also reduce the chances of race conditions with getting the unloaded cogs/loading cogs/checking for not found cogs.

File: bot/bot.py
Comment (Shivansh-007) [reply bot/bot.py:50]: Yeah these are for local testing, I have hence left a comment over that and in the PR description. I will remove that before it gets merged.

File: bot/bot.py
Comment (Shivansh-007) [reply bot/bot.py:44]: > Is this guaranteed to run after all the extensions have been loaded?

It should, yes, and while testing it did.

Comment (ToxicKidz) [reply bot/bot.py:44]: Yeah, the event loop gets ran after the bot instance is created

**Comment by Shivansh-007** [comment]
> With your current approach, you store the cached unloads into a dict, then remove successfully unloaded cogs from the dict in `load_extension`, then compare the dict with the cached unloads to find missing cogs.
> 
> A better solution would be to just go ahead with loading the extension (and checking presence in the cache before loading), and storing a set of successfully unloaded cogs instead. Then, you can get all the unloads from the cache and clean up the ones that were not in the successful dict.
> 
> This can be done with a `load_extensions` method on the bot that keep tracks of extensions that were successfully unloaded locally. That way, there's no need to store a redundant instance attribute after startup. It would also reduce the chances of race conditions with getting the unloaded cogs/loading cogs/checking for not found cogs.
 
That would be possible if we make `load_extensions` an async function, since to check whether it exists within the async redis-cache or not it would be awaited. Hence I didn't do that.

**Comment by kosayoda** [comment]
Would that be an issue? It wouldn't be much unlike how you're getting the unloaded extensions after all. I'm not an async expert though

**Comment by Shivansh-007** [comment]
> Would that be an issue? It wouldn't be much unlike how you're getting the unloaded extensions after all. I'm not an async expert though

Not really but I would need to add `await`/`async` in the some places (The extensions cog, and `bot/__main__.py` probably).

File: bot/__main__.py
Comment (brad90four) [inline bot/__main__.py:35]:
```py
    await bot.unloads_cache.set("bot.exts.evergreen.catify", "https://dummy-msg-link.com")
    await bot.unloads_cache.set("bot.exts.evergreen.xkd", "https://dummy-msg-link.com")
```

Are these lines the "testing" code that you mentioned?

File: bot/exts/utils/extensions.py
Comment (brad90four) [inline bot/exts/utils/extensions.py:231]:
```suggestion
    async def manage(self, action: Action, ext: str) -> tuple[str, t.Optional[str]]:
```

**Review by brad90four — COMMENTED** [review]
Just giving it a read through for the content. I haven't set it up or tested it out locally and I only saw two things to comment on. The first one is potentially the "testing" code as a place holder, and the other is related to the fix type annotations PR that is being worked on.

File: bot/__main__.py
Comment (Shivansh-007) [reply bot/__main__.py:35]: Yes.

**Comment by Xithrius** [comment]
File restructure has occurred, it may have been the cause of these conflicts.

File: bot/bot.py
Comment (Bluenix2) [inline bot/bot.py:89]: Shouldn't you also override `unload_extension()` to add the extension to the unloaded cache?

File: bot/bot.py
Comment (Bluenix2) [inline bot/bot.py:226]: What is this link?

File: bot/bot.py
Comment (Bluenix2) [inline bot/bot.py:235]: This does not sucessfully handle the case of someone having bad configuration and these returning None.

File: bot/bot.py
Comment (Bluenix2) [inline bot/bot.py:237]: What is `\u26a0`?

Can you use the [`\N{NAME}`](https://docs.python.org/3/howto/unicode.html#the-string-type) style?

File: bot/bot.py
Comment (Bluenix2) [inline bot/bot.py:244]: What happens if you delete half of these then one of them fail? You won't send the message that you have cleared the cache because it errors. *That said it will send a sentry error and appear in the log I believe, but my point still stands*

Can you move this down until after you have sent the message?

**Review by Bluenix2 — CHANGES_REQUESTED** [review]

File: bot/bot.py
Comment (Bluenix2) [reply bot/bot.py:97]: Why is the default False?

Comment (Shivansh-007) [reply bot/bot.py:97]: Since we load the cog earlier and then remove them from the cache (when using the `.c load` command, it would never be loaded as the cog would be present in the cache, so this option was added to force load the cog without checking the cache.

Since we don't want to force reload the cog when the bot initially loads the cogs, it is set to `False`.

Comment (Bluenix2) [reply bot/bot.py:97]: But this means that when `start()` calls `load_extension()` it will always load all cogs?

Because you do `if ignore_cache and not await self.unloads_cache.get(name):`?

I am a bit confused

Comment (Shivansh-007) [reply bot/bot.py:97]: Ah yeah, I had fixed that locally during my last testing... forgot to push it. Should be fixed now in 8193125.

File: bot/bot.py
Comment (Bluenix2) [inline bot/bot.py:102]:
```suggestion
            log.info(f"Skipping cog {name}, found in unloaded cache.")
```
```suggestion
            log.info(f"Skipping ext '{name}', found in unloaded cache.")
```

File: bot/bot.py
Comment (Bluenix2) [inline bot/bot.py:100]: This if-statement is still wrong, but now for the wrong reason. `await self.unloads_cache.get(name)` becomes None and then you invert this using `not`.

What ends up happening is the opposite of what we want.

**Review by Bluenix2 — CHANGES_REQUESTED** [review]

File: bot/bot.py
Comment (Bluenix2) [inline bot/bot.py:78]: Can this be removed please, as a rule of thumb unless a PR is marked as a Draft or has requested changes you should feel comfortable having it merged at any time.

Comment (Shivansh-007) [reply bot/bot.py:78]: Yes, I thought I removed it from my commit, was testing... so needed that.

**Review by Bluenix2 — APPROVED** [review]
Let's get this merged! ✨ Thanks for this @Shivansh-007

File: bot/bot.py
Comment (onerandomusername) [inline bot/bot.py:75]: why is this imported here? is it possible to import it at the top of the file?

File: bot/bot.py
Comment (onerandomusername) [inline bot/bot.py:155]: good change imo

File: bot/bot.py
Comment (onerandomusername) [reply bot/bot.py:250]: This may be worth not assuming, given that a user testing pulls may come across this error more than the referenced one.

File: bot/constants.py
Comment (onerandomusername) [inline bot/constants.py:111]: `CHANNEL_DEV_ALERTS`?

**Review by onerandomusername — CHANGES_REQUESTED** [review]
There is somewhat a problem here, conceptually.

Internally on dpy, load_extension, unload_extension, and reload_extension are all *not* awaitables.

However, the new load_extension for the bot is defined as a coro. There may be internal methods that call load_extension, but I don't think there are, so it should be good.

However, because of these three lines, 

https://github.com/python-discord/sir-lancebot/blob/029a6324823d4d8e934c12edbc41d221e59d809a/bot/exts/core/extensions.py#L24-L30

These need to all be awaited too. The problem arises in unload_extension and reload_extension. Having not been overridden, they are still normal functions, not coros. As reading the code, I can see that they are all awaited where they are used, but the functions themselves are still normal functions, and not coros.

File: bot/bot.py
Comment (Shivansh-007) [reply bot/bot.py:75]: Circular imports.

File: bot/exts/core/extensions.py
Comment (onerandomusername) [inline bot/exts/core/extensions.py:241]: I may have missed this code, ignore my previous write-up

**Review by onerandomusername — APPROVED** [review]
er, lgtm

**Comment by Xithrius** [comment]
Per [this message](https://canary.discord.com/channels/267624335836053506/635950537262759947/940807906264219678) from the author, this PR is now up for grabs.

**Comment by Xithrius** [comment]
This PR is now up for grabs. Would anyone that's currently subscribed to this thread or just happened to see this comment like to take over this feature?

**Comment by ToxicKidz** [comment]
Hello, I’d like to have a go at this in a separate PR.

**Comment by Shivansh-007** [comment]
@Xithrius Currently this PR is complete, only needs another core-dev approval.

**Comment by Shivansh-007** [comment]
Cool okay 👌🏻 , let me know if you need any help while going through the code @ToxicKidz.

**Comment by Xithrius** [comment]
Whoops, didn't mean to request another review from you, @Bluenix2. Sorry about that.

**Comment by Xithrius** [comment]
@ToxicKidz When you have the time, please resolve the file conflicts within this PR. Thanks!

**Comment by Xithrius** [comment]
Due to the staleness of this PR, I'll be closing it. If anyone wants to pick it back up, feel free to do so.

**Pull request closed without merging**
